### PR TITLE
Add SDL_GetDefaultAudioInfo

### LINKED ISSUE #5866: audio: Add GetDefaultAudioSpec
With a very specific function signature:

```
int SDL_GetDefaultAudioSpec(SDL_AudioSpec *spec, int iscapture);
```

This is meant to be an answer for the fact that GetAudioDeviceSpec can _not_ support this idea, for the various reasons that have been brought up in previous discussions. If the backend has a way for us to ask explicitly for a default endpoint it will fill `spec`, but if not (either because it doesn't have an API or the API does not specify a default) it returns `SDL_Unsupported()`. While the device list can be separate from what the default will be, backends _do_ have separate calls for the mystical `NULL` case:

- Windows has [IMMDeviceEnumerator_GetDefaultAudioEndPoint](https://docs.microsoft.com/en-us/windows/win32/api/mmdeviceapi/nf-mmdeviceapi-immdeviceenumerator-getdefaultaudioendpoint)
- PulseAudio's [`pa_server_info`](https://freedesktop.org/software/pulseaudio/doxygen/structpa__server__info.html) has `default_sink_name` and `default_source_name` properties
- PipeWire is already implemented in SDL, under the `metadata_property` callback

I'd have to look it up for CoreAudio, but I'd be surprised if it didn't have something for this.

UPDATE: Typical PipeWire, it's already implemented in our code for it!

### PR DESCRIPTION
**Pull request by flibitijibibo**
Currently this is supported on pipewire, pulseaudio, and wasapi.

~~This is flagrantly untested!~~ Also, I haven't looked at CoreAudio, but I probably won't since I'm pretty far removed from Apple stuff at this point.

Fixes #5866

### CODE REVIEW

**Review by Kontrabant — COMMENTED** [review]
Note that the Pipewire default values changed from ints to JSON strings in 0.3.24 and SDL supports as far back as 0.3.20.  To avoid potential breakage this either needs to select the int/string behavior via a version check, or the minimum version needs to be bumped.

**Comment by flibitijibibo** [comment]
Latest push has been tested on all three backends. WASAPI is fine, PipeWire and Pulse involved a lot of annoying work to link the ALSA identifier with the descriptions we use for the device name. Also, PipeWire's identifier is wrapped in JSON for some reason.

**Comment by flibitijibibo** [comment]
> Note that the Pipewire default values changed from ints to strings in 0.3.24 and SDL supports as far back as 0.3.20. To avoid potential breakage this either needs to select the int/string behavior via a version check, or the minimum version needs to be bumped.



> Note that the Pipewire default values changed from ints to strings in 0.3.24 and SDL supports as far back as 0.3.20. To avoid potential breakage this either needs to select the int/string behavior via a version check, or the minimum version needs to be bumped.

It's looking like we might bump the version anyhow, probably to line up with the init work for #5268, unless we end up getting a new function call for session checks.

File: src/audio/pulseaudio/SDL_pulseaudio.c
Comment (sezero) [inline src/audio/pulseaudio/SDL_pulseaudio.c:849]: intptr_t should be preferred here instead of size_t

**Comment by Kontrabant** [comment]
> It's looking like we might bump the version anyhow, probably to line up with the init work for #5268, unless we end up getting a new function call for session checks.

That's fine then.  I don't think it will be a problem as anything that actually uses Pipewire for audio is generally on a much newer version. Even the current Ubuntu LTS is on 0.3.48 now.

File: src/audio/pulseaudio/SDL_pulseaudio.c
Comment (flibitijibibo) [reply src/audio/pulseaudio/SDL_pulseaudio.c]: Done!

**Review by Kontrabant — CHANGES_REQUESTED** [review]

File: src/audio/pulseaudio/SDL_pulseaudio.c
Comment (sezero) [inline src/audio/pulseaudio/SDL_pulseaudio.c:776]: This (and its likes) too need changing from size_t to (u)intptr_t I guess. (although this one seems to have been here for some time.)

Comment (flibitijibibo) [reply src/audio/pulseaudio/SDL_pulseaudio.c:776]: I'll do a sweep through pulseaudio.c in a separate PR; for now I've got intptr_t in all the new lines at least!

**Comment by flibitijibibo** [comment]
One last push to clean up the documentation. At this point I'm happy with this function, and it fulfills a need that FAudio, FMOD_SDL, and Wwise_SDL have wanted for _years_ now. I'd still like a review from at least one more maintainer though, as this does dig pretty deep into the enumeration code for PulseAudio in particular. (WASAPI and PipeWire were pretty noninvasive, so those should be fine.)

**Comment by flibitijibibo** [comment]
This will conflict with #5750, but one doesn't necessarily block the other... really all that's different is that DirectSound will be able to support this function as well.

**Comment by flibitijibibo** [comment]
Largely a repost of #5902: I looked at the milestone and it's exactly two weeks from today, and we haven't pushed a pre-release in about three weeks, so I'm going to try the Organ Donor approach: If nobody says _no_ to this patch I'll merge tomorrow morning, since this is a super isolated feature and is effectively a replica of two APIs we already have. Just about the only thing we could _add_ is channel masks, but that's something that would be more appropriate in 3.0 since that (at least IMO) is better in the AudioSpec struct.

**Comment by flibitijibibo** [comment]
Rebased to include directsound support!

**Comment by icculus** [comment]
> Can you confirm that we want to add this? It seems like a good idea to me, but wanted to double check before this gets merged.

It touches a lot of code (which is a necessity here), but I think it's okay. I think we're likely to miss the milestone deadline this time, so it'll have some time to marinate before we ship if we merge it now.

**Comment by flibitijibibo** [comment]
Makes sense to me - if anyone is interested in testing this, the two things I would prioritize over anything else is the PulseAudio "change" event - it's kind of vague so it's hard to know the full extent of when those happen, but at least in my stress testing it was only ever fired in order to notify default device changes or format changes, the latter of which was never supported anyway (and ultimately those events fall through without touching anything).

I should be around to address any cases where I'm wrong; this is FNA's #1 priority even over GDK and Wayland!

**Pull request merged**
